aws/get_running_instances.py

```diff
@@ -10,18 +10,59 @@
     Else, you can add the aws keys as arguments to the above command:
         ** '--aws-access-key-id' or '-i'
         ** '--aws-secret-access-key' or '-s'
+
+Example Output:
+    2015-05-28 16:59:02,421 [INFO] datasrc=aws, \
+    jenkins_master=example.jenkins.com, total_executors_jenkins=86, \
+    total_executors_ec2=86, desc-1-worker_count=5, desc-2-worker_count=81
+
+Assumptions:
+    This is implemented with some assumptions about the way that workers are
+    tagged. In particuar, it is assumed that the tags `master` and `worker` are
+    used. Values for `master` are expected to be the netloc of the jenkins
+    master.  Values for `worker` are expected to match the description given
+    in the Jenkins EC2 plugin configuration, but without ending in '-worker'.
+    For example, if my worker description in jenkins is 'abc-worker', then
+    the worker tag in EC2 is expected to be 'abc'.
 """
 import argparse
 import boto
 import logging
 import os
 import sys
+from collections import Counter, defaultdict
 
-
-logging.basicConfig(format='[%(levelname)s] %(message)s')
 logger = logging.getLogger(__name__)
 
 
+def sort_by_master_and_worker(reservations, key_name):
+    """
+    Sorts the list of boto `Reservation` objects into a dictionary,
+    mapping `master` tags to a counter of `worker` tags for that master.
+    Only running and pending instances matching the given `key_name` will be
+    included. Instances without `master` or `worker` tags will be sorted as
+    'untagged'.
+
+    :Args:
+        reservations: a list of boto `Reservation` objects
+        key_name: an ec2 key_name to match
```

Comment:
We can tell orphaned workers from other non-tagged workers that way.  The orphaned workers have the key_name, but the master and worker tags are missing.  We can't just look at the tags because there are workers with other key_names that aren't tagged and aren't orphaned.